lms/djangoapps/course_api/blocks/transformers/tests/test_milestones.py

```diff
@@ -0,0 +1,280 @@
+                'is_practice_exam': False,
+                '#children': [
+                    {'#type': 'vertical', '#ref': 'J'},
+                    {'#type': 'vertical', '#ref': 'K'},
+                ],
+            },
+        ]
+
+    def test_exam_not_created(self):
+        self.get_blocks_and_check_against_expected(self.user, self.ALL_BLOCKS_EXCEPT_TIMED)
+
+    @ddt.data(
+        (
+            'ProctoredExam',
+            ProctoredExamStudentAttemptStatus.declined,
+            ALL_BLOCKS_EXCEPT_TIMED
+        ),
+        (
+            'ProctoredExam',
+            ProctoredExamStudentAttemptStatus.submitted,
+            ('course', 'A', 'B', 'C', 'PracticeExam', 'F', 'G', 'H', 'I'),
+        ),
+        (
+            'ProctoredExam',
+            ProctoredExamStudentAttemptStatus.rejected,
+            ('course', 'A', 'B', 'C', 'PracticeExam', 'F', 'G', 'H', 'I'),
+        ),
+        (
+            'PracticeExam',
+            ProctoredExamStudentAttemptStatus.declined,
+            ALL_BLOCKS_EXCEPT_TIMED
+        ),
+        (
+            'PracticeExam',
+            ProctoredExamStudentAttemptStatus.rejected,
+            ('course', 'A', 'B', 'C', 'ProctoredExam', 'D', 'E', 'H', 'I'),
+        ),
+    )
+    @ddt.unpack
+    def test_exam_created(self, exam_ref, attempt_status, expected_blocks):
+        self.setup_proctored_exam(self.blocks[exam_ref], attempt_status, self.user.id)
+        self.get_blocks_and_check_against_expected(self.user, expected_blocks)
+
+    @ddt.data(
+        (
+            'PracticeExam',
+            'ProctoredExam',
+            'D',
+            ('course', 'A', 'B', 'C', 'ProctoredExam', 'D', 'E', 'H', 'I')
+        ),
+        (
+            'H',
+            'ProctoredExam',
+            'D',
+            ('course', 'A', 'B', 'C', 'ProctoredExam', 'D', 'E', 'PracticeExam', 'F', 'G'),
+        ),
+    )
+    @ddt.unpack
+    def test_gated(self, gated_block_ref, gating_block_ref, gating_block_child, expected_blocks_before_completion):
+        """
+        First, checks that a student cannot see the gated block when it is gated by the gating block and no
+        attempt has been made to complete the gating block.
+        Then, checks that the student can see the gated block after the gating block has been completed.
+
+        expected_blocks_before_completion is the set of blocks we expect to be visible to the student
+        before the student has completed the gating block.
+
+        The test data includes one special exam and one non-special block.
+        """
+        self.course.enable_subsection_gating = True
+        self.setup_gated_section(self.blocks[gated_block_ref], self.blocks[gating_block_ref])
+        self.get_blocks_and_check_against_expected(self.user, expected_blocks_before_completion)
+
+        # mock the api that the lms gating api calls to get the score for each block to always return 1 (ie 100%)
+        import courseware
```

Comment:
Unless there's a circular reference for some reason, all imports should be at the top of the file.